drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1453 @@
+{
+    FUNCTION_TRACE;
+
+    return WskStreamIo(Socket, Buffer, Flags, Irp, DIR_RECEIVE);
+}
+
+static NTSTATUS WSKAPI
+WskDisconnect(_In_ PWSK_SOCKET Socket, _In_opt_ PWSK_BUF Buffer, _In_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    FUNCTION_TRACE;
+
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+
+static WSK_PROVIDER_CONNECTION_DISPATCH TcpDispatch = {
+    .WskControlSocket = WskControlSocket,
+    .WskCloseSocket = WskCloseSocket,
+    .WskBind = WskBind,
+    .WskConnect = WskConnect,
+    .WskGetLocalAddress = WskGetLocalAddress,
+    .WskGetRemoteAddress = WskGetRemoteAddress,
+    .WskSend = WskSend,
+    .WskReceive = WskReceive,
+    .WskDisconnect = WskDisconnect,
+    .WskRelease = WskReleaseTcp,
+};
+
+static NTSTATUS WSKAPI
+WskSocket(
+    _In_ PWSK_CLIENT Client,
+    _In_ ADDRESS_FAMILY AddressFamily,
+    _In_ USHORT SocketType,
+    _In_ ULONG Protocol,
+    _In_ ULONG Flags,
+    _In_opt_ PVOID SocketContext,
+    _In_opt_ const VOID * Dispatch,
+    _In_opt_ PEPROCESS OwningProcess,
+    _In_opt_ PETHREAD OwningThread,
+    _In_opt_ PSECURITY_DESCRIPTOR SecurityDescriptor,
+    _Inout_ PIRP Irp)
+{
+    PWSK_SOCKET_INTERNAL s;
+    NTSTATUS status;
+
+    FUNCTION_TRACE;
+
+    IoSetNextIrpStackLocation(Irp);
+
+    if (AddressFamily != AF_INET)
+    {
+        DPRINT1("Address family %d not supported\n", AddressFamily);
+        status = STATUS_NOT_SUPPORTED;
+        goto err_out;
+    }
+    switch (SocketType)
+    {
+        case SOCK_DGRAM:
+            if (Protocol != IPPROTO_UDP)
+            {
+                DPRINT1("SOCK_DGRAM only supports IPPROTO_UDP\n");
+                status = STATUS_INVALID_PARAMETER;
+                goto err_out;
+            }
+            if (Flags != WSK_FLAG_DATAGRAM_SOCKET)
+            {
+                DPRINT1("SOCK_DGRAM flags must be WSK_FLAG_DATAGRAM_SOCKET\n");
+                status = STATUS_INVALID_PARAMETER;
+                goto err_out;
+            }
+            break;
+
+        case SOCK_STREAM:
+            if (Protocol != IPPROTO_TCP)
+            {
+                DPRINT1("SOCK_STREAM only supports IPPROTO_TCP\n");
+                status = STATUS_INVALID_PARAMETER;
+                goto err_out;
+            }
+            if ((Flags != WSK_FLAG_CONNECTION_SOCKET) && (Flags != WSK_FLAG_LISTEN_SOCKET))
+            {
+                DPRINT1("SOCK_STREAM flags must be either WSK_FLAG_CONNECTION_SOCKET or WSK_FLAG_LISTEN_SOCKET\n");
+                status = STATUS_INVALID_PARAMETER;
+                goto err_out;
+            }
+            break;
+
+        case SOCK_RAW:
+            DPRINT1("SOCK_RAW not supported\n");
+            status = STATUS_NOT_SUPPORTED;
+            goto err_out;
+
+        default:
+            status = STATUS_INVALID_PARAMETER;
+            goto err_out;
+    }
+
+    s = ExAllocatePoolWithTag(NonPagedPool, sizeof(*s), TAG_NETIO);
```

Comment:
```suggestion
    s = ExAllocatePoolZero(NonPagedPool, sizeof(*s), TAG_NETIO);
```
would get rid of the manual initialization below.

Reply:
Ok, thanks applied.